influx-reports/influx.py

```diff
@@ -0,0 +1,241 @@
+
+    def __init__(self, output_path):
+        self.output_path = output_path
+
+    def __enter__(self):
+        self.output_stream = open(self.output_path, 'wb+')
+        self.output = StreamResultToBytes(self.output_stream)
+        self.output.startTestRun()
+        return self.output
+
+    def __exit__(self, *args, **kwargs):
+        self.output.stopTestRun()
+        self.output_stream.close()
+
+    def status(self, *args, **kwargs):
+        self.output.status(*args, **kwargs)
+
+
+def return_time(client, query, delta_seconds=0):
+    """ From an InfluxDB query, fetch
+    the first point time, and return a
+    python time object. Shift it from
+    a few seconds (delta_seconds) if necessary.
+    """
+    timestamp_query = client.query(query)
+    # Get points is generator, we should just get first
+    # point time (string type).
+    fpt_str = next(timestamp_query.get_points())['time']
+    try:
+        fpt = dateutil.parser.parse(
+            fpt_str) + datetime.timedelta(seconds=delta_seconds)
+    except Exception as alle:
+        raise InfluxTimestampParseException(
+            "Error parsing a timestamp from influx: {}".format(alle))
+    return fpt
+
+
+def calculate_measurement_uptime(client, measurement, job_reference,
+                                 measurement_period_seconds,
+                                 resolution=60):
+    """ For a certain build (job_reference), estimate the amount
+    of seconds a component (column of table) was up.
+    For that, we slice the job by resolution.
+    If I have no data in the slice, I assume downtime. If I have
+    at least one data per slice, I assume up, even if degraded.
+    The count of these positive events multiplied by resolution
+    is an average idea of the uptime.
+    """
+    per_field_up_percent=defaultdict(float)
+    measurement_uptime_percent = 0.0
+
+    query = (
+        "select max(/.*_status/) from {measurement} "
+        "where time < now() and job_reference='{job_reference}' "
+        "group by time({resolution}s) fill(-1)"
+    ).format(measurement=measurement,
+             job_reference=job_reference,
+             resolution=resolution)
+
+    max_amount_of_time_slices = math.ceil(float(measurement_period_seconds)/resolution)
+
+    all_data = client.query(query)
+    for time_slice in all_data.get_points():
+        measurement_was_up = False
+        for element in time_slice:
+            if element != 'time' and time_slice[element] == 1:
+                key = element.replace('max_', '').replace('_status','')
+                # Mark the field as up
+                per_field_up_percent[key] += (1 / max_amount_of_time_slices)
+                # Mark the measurement as up
+                measurement_was_up = True
+        if measurement_was_up:
+            measurement_uptime_percent += (1 / max_amount_of_time_slices)
+
+    return dict(
+        measurement_uptime_percent=measurement_uptime_percent,
```

Comment:
The mean uptime of a services components does not give any useful measure of the availability of the service, I think it should be removed.